velox/functions/lib/aggregates/MinMaxByAggregatesBase.h

```diff
@@ -31,631 +34,297 @@ constexpr bool isNumeric() {
-/// takes a pair of X and min/max Y and returns X. T is the type of X and U is
-/// the type of Y.
 template <
-    typename T,
-    typename U,
+    typename V,
+    typename C,
     bool isMaxFunc,
-    template <bool B, typename C1, typename C2>
+    template <bool C0, typename C1, typename C2>
     class Comparator>
-class MinMaxByAggregateBase : public exec::Aggregate {
+class MinMaxByAggregate {
  public:
+  using InputType = Row<Generic<T1>, Orderable<T2>>;
+  using IntermediateType = Row<Generic<T1>, Orderable<T2>>;
+  using OutputType = Generic<T1>;
+
   using ValueAccumulatorType =
-      typename AccumulatorTypeTraits<T>::AccumulatorType;
+      typename AccumulatorInternalTypeTraits<V>::AccumulatorInternalType;
   using ComparisonAccumulatorType =
-      typename AccumulatorTypeTraits<U>::AccumulatorType;
+      typename AccumulatorInternalTypeTraits<C>::AccumulatorInternalType;
 
-  explicit MinMaxByAggregateBase(
-      TypePtr resultType,
-      bool throwOnNestedNulls = false)
-      : exec::Aggregate(resultType), throwOnNestedNulls_(throwOnNestedNulls) {}
+  // Default null behavior is false because for both spark sql and presto sql,
+  // max/min_by can still be done if value is null
+  static constexpr bool default_null_behavior_ = false;
 
-  int32_t accumulatorFixedWidthSize() const override {
-    return sizeof(ValueAccumulatorType) + sizeof(ComparisonAccumulatorType) +
-        sizeof(bool);
-  }
+  struct AccumulatorType {
+    std::optional<ValueAccumulatorType> currValue_;
+    std::optional<ComparisonAccumulatorType> currComparison_;
 
-  void addRawInput(
-      char** groups,
-      const SelectivityVector& rows,
-      const std::vector<VectorPtr>& args,
-      bool /*unused*/) override {
-    addRawInput(
-        groups,
-        rows,
-        args,
-        [&](auto* accumulator,
-            const auto& newComparisons,
-            auto index,
-            auto isFirstValue) {
-          return Comparator<isMaxFunc, U, ComparisonAccumulatorType>::compare(
-              accumulator, newComparisons, index, isFirstValue);
-        });
-  }
+    explicit AccumulatorType(HashStringAllocator* /*allocator*/) {}
 
-  void addIntermediateResults(
-      char** groups,
-      const SelectivityVector& rows,
-      const std::vector<VectorPtr>& args,
-      bool /*mayPushdown*/) override {
-    addIntermediateResults(
-        groups,
-        rows,
-        args,
-        [&](auto* accumulator,
-            const auto& newComparisons,
-            auto index,
-            auto isFirstValue) {
-          return Comparator<isMaxFunc, U, ComparisonAccumulatorType>::compare(
-              accumulator, newComparisons, index, isFirstValue);
-        });
-  }
+    bool addInput(
+        HashStringAllocator* allocator,
+        exec::optional_arg_type<Generic<T1>> value,
+        exec::optional_arg_type<Orderable<T2>> comparison) {
+      // Input will be ignored if comparison is null
+      if (!comparison.has_value()) {
```

Comment:
I think the check may not be necessary since we allow value to be null? The `update` function in line 213 has logic to specifically deal with the case when value is null